### PR TITLE
[tabular] Fix balanced_accuracy edge-case

### LINKED ISSUE #4590: [BUG] TabularPredictor: Validation raises an exception for a 2-class string target column
**Bug Report Checklist**

- [x] I provided code that demonstrates a minimal reproducible example.  
- [ ] I confirmed bug exists on the latest mainline of AutoGluon via source install.  
- [x] I confirmed bug exists on the latest stable version of AutoGluon.  

**Describe the bug**
When a TabularPredictor is trained on a 2-class problem in "multiclass" mode and then evaluated, the evaluation fails with an autogluon-internal exception during balanced accuracy calculation (see below). This especially happens for a target column which is of type string.

**Expected behavior**
A 2-class TabularPredictor in "multiclass" mode can be trained and evaluated without exceptions, just as a 3-class or 4-class prediction.

**To Reproduce**
```
from autogluon.tabular import TabularPredictor
from autogluon.core.metrics.classification_metrics import balanced_accuracy
import numpy as np
import pandas as pd

x1 = [1,2,1,2,3,4,3,2,1,2,3,4,3,2,3,4,5,3,2,1]
x2 = [4,3,2,3,2,1,4,5,6,3,7,3,2,8,4,2,3,5,4,1]
solution = np.array(["ABC","DEF","ABC","ABC","DEF","ABC","DEF","ABC","ABC","DEF","ABC","DEF","ABC","ABC","DEF","ABC","DEF","ABC","ABC","DEF"])
full_data = pd.DataFrame({"x1":x1, "x2":x2, "label":solution})
model = TabularPredictor("label", problem_type="multiclass")
model.fit(full_data, time_limit=60)

probas = model.predict_proba(full_data)
eval_res = model.evaluate_predictions(full_data["label"], probas, detailed_report=True)  # THIS CRASHES
```

Reproduction of the actual library function call:

```
from autogluon.tabular import TabularPredictor
from autogluon.core.metrics.classification_metrics import balanced_accuracy
import numpy as np
import pandas as pd

# 3 classes: works
solution = np.array(["ABC","DEF","ABC","ABC","DEF","ABC","DEF","ABC","ABC","DEF","ABC","DEF","ABC","ABC","DEF","ABC","DEF","ABC","ABC","DEF","GHI"])
prediction = np.array(["ABC","DEF","DEF","ABC","DEF","ABC","DEF","ABC","DEF","ABC","ABC","DEF","ABC","ABC","ABC","ABC","DEF","ABC","ABC","DEF","GHI"])
print(balanced_accuracy(solution, prediction))

# 2 classes: crashes
solution = np.array(["ABC","DEF","ABC","ABC","DEF","ABC","DEF","ABC","ABC","DEF","ABC","DEF","ABC","ABC","DEF","ABC","DEF","ABC","ABC","DEF"])
prediction = np.array(["ABC","DEF","DEF","ABC","DEF","ABC","DEF","ABC","DEF","ABC","ABC","DEF","ABC","ABC","ABC","ABC","DEF","ABC","ABC","DEF"])
print(balanced_accuracy(solution, prediction))  # THIS CRASHES
```

**Screenshots / Logs**
Here is the exception stack trace excerpt of the full training example:
```
  File "/data/src/src/test.py", line 25, in <module>
    eval_res = model.evaluate_predictions(full_data["label"], probas, detailed_report=True)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/python/.local/lib/python3.11/site-packages/autogluon/tabular/predictor/predictor.py", line 2330, in evaluate_predictions
    return self._learner.evaluate_predictions(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/python/.local/lib/python3.11/site-packages/autogluon/tabular/learner/abstract_learner.py", line 786, in evaluate_predictions
    score = self._score_with_pred_proba(
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/python/.local/lib/python3.11/site-packages/autogluon/tabular/learner/abstract_learner.py", line 658, in _score_with_pred_proba
    return compute_weighted_metric(y_tmp, y_pred, metric, weights=sample_weight, weight_evaluation=weight_evaluation, quantile_levels=self.quantile_levels)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/python/.local/lib/python3.11/site-packages/autogluon/core/utils/utils.py", line 673, in compute_weighted_metric
    return metric(y, y_pred, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/python/.local/lib/python3.11/site-packages/autogluon/core/metrics/__init__.py", line 96, in __call__
    return self._score(y_true=y_true, y_pred=y_pred, **k)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/python/.local/lib/python3.11/site-packages/autogluon/core/metrics/__init__.py", line 134, in _score
    return self._sign * self._score_func(y_true, y_pred, **kwargs)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/python/.local/lib/python3.11/site-packages/autogluon/core/metrics/classification_metrics.py", line 51, in balanced_accuracy
    fn = np.sum(np.multiply(solution, (1 - prediction)), axis=0, dtype=float)
                                       ~~^~~~~~~~~~~~
TypeError: unsupported operand type(s) for -: 'int' and 'str'
```

**Installed Versions**

<details>

```INSTALLED VERSIONS
------------------
date                : 2024-10-28
time                : 14:24:55.361991
python              : 3.11.10.final.0
OS                  : Linux
OS-release          : 5.15.153.1-microsoft-standard-WSL2
Version             : #1 SMP Fri Mar 29 23:14:13 UTC 2024
machine             : x86_64
processor           : 
num_cores           : 8
cpu_ram_mb          : 7872.03125
cuda version        : None
num_gpus            : 0
gpu_ram_mb          : []
avail_disk_size_mb  : 120097

autogluon           : None
autogluon.common    : 1.1.1
autogluon.core      : 1.1.1
autogluon.features  : 1.1.1
autogluon.tabular   : 1.1.1
boto3               : 1.35.41
catboost            : 1.2.7
fastai              : None
hyperopt            : 0.2.7
imodels             : None
lightgbm            : 4.3.0
matplotlib          : 3.9.2
networkx            : 3.4.1
numpy               : 1.26.4
onnxruntime-gpu     : None
pandas              : 2.2.3
psutil              : 5.9.8
ray                 : 2.10.0
requests            : 2.32.3
scikit-learn        : 1.4.0
scikit-learn-intelex: None
scipy               : 1.12.0
setuptools          : 65.5.1
skl2onnx            : None
tabpfn              : None
torch               : 2.3.1+cpu
tqdm                : 4.66.5
vowpalwabbit        : None
xgboost             : 2.0.3
```

</details>

Thank you!

### PR DESCRIPTION
**Pull request by Innixma**
*Issue #, if available:*

Resolves #4590 
Closes #4750

*Description of changes:*

Fix balanced_accuracy edge-case exception detailed in #4590.

Additionally adds many metric unit tests to better detect bugs like this in future. The new unit tests catch the previous bug.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: core/tests/unittests/metrics/test_metrics.py [hunk 1]
```diff
@@ -140,6 +152,60 @@ def test_metrics_perfect_multiclass(metric: str):
     _assert_perfect_score(scorer=METRICS[MULTICLASS][metric])
 
 
+@pytest.mark.skip(reason="average_precision doesn't raise an exception here when it should, so this test currently fails")
```
Comment (suzhoum) [inline core/tests/unittests/metrics/test_metrics.py:155]: Do we know why it fails?

File: core/tests/unittests/metrics/test_metrics.py [hunk 2]
```diff
@@ -140,6 +152,60 @@ def test_metrics_perfect_multiclass(metric: str):
...
+@pytest.mark.parametrize("metric", METRICS_NEEDS_THRESHOLD[BINARY], ids=METRICS_NEEDS_THRESHOLD[BINARY])  # noqa
+def test_metrics_perfect_raises_binary_single_sample(metric: str):
+    with pytest.raises(Exception):
+        # threshold metrics should not be able to predict on a single sample
+        _assert_perfect_score_single_sample(scorer=METRICS[BINARY][metric])
+
+
+@pytest.mark.skip(reason="mcc, quadradic_kappa, and pac produce unexpected values here, so this test currently fails")
```
Comment (suzhoum) [inline core/tests/unittests/metrics/test_metrics.py:163]: Do we plan to fix these tests in a follow up PR?

File: core/src/autogluon/core/metrics/classification_metrics.py [hunk 3]
```diff
@@ -25,7 +25,13 @@ def balanced_accuracy(solution, prediction):
 
     if y_type == "binary":
         # Do not transform into any multiclass representation
-        pass
+        unique_sol = np.unique(solution)
+        unique_pred = np.unique(prediction)
+        classes = np.unique(np.concatenate((unique_sol, unique_pred)))
+        if set(classes) != {0, 1}:
+            pos_class = classes[-1]
```
[the remark this answers is not in the document]
Comment (Innixma) [reply core/src/autogluon/core/metrics/classification_metrics.py:32]: `max(classes)` will raise an exception if the values are strings, which they can be. In the balanced_accuracy metric which class is the positive class doesn't matter, since they have equal weight, so the selection mechanism for `pos_class` is fine as long as it is consistent.

File: core/tests/unittests/metrics/test_metrics.py (hunk 1 above)
Comment (Innixma) [reply core/tests/unittests/metrics/test_metrics.py:155]: This is an edge-case where the user tries to score a single sample's prediction.

It is the output of the scikit-learn metric, so it isn't a bug caused by AutoGluon's wrapper. It isn't a new issue, the new tests simply showed that this failure case exists.

It is likely because average precision (probably?) should be undefined for a single sample. https://scikit-learn.org/1.5/modules/model_evaluation.html#precision-recall-f-measure-metrics

However, sklearn decided instead of raising an exception to return a score of 1.0 if y_true = [1], ignoring the prediction (y_score), and similarly return a score of 0.0 if y_true != [1], ignoring the prediction.

```python
from sklearn.metrics import average_precision_score

y_score = [0]
y_true = [1]
score = average_precision_score(y_true=y_true, y_score=y_score)
print(score)  # 1.0

y_score = [1]
y_true = [1]
score = average_precision_score(y_true=y_true, y_score=y_score)
print(score)  # 1.0

y_score = [0.5]
y_true = [1]
score = average_precision_score(y_true=y_true, y_score=y_score)
print(score)  # 1.0

y_score = [0]
y_true = [0]
score = average_precision_score(y_true=y_true, y_score=y_score)
print(score)  # 0.0

y_score = [1]
y_true = [0]
score = average_precision_score(y_true=y_true, y_score=y_score)
print(score)  # 0.0

y_score = [0.5]
y_true = [0]
score = average_precision_score(y_true=y_true, y_score=y_score)
print(score)  # 0.0
```

Comment (Innixma) [reply core/tests/unittests/metrics/test_metrics.py:155]: Because it is a threshold metric, and my understanding is that threshold metrics should fail if used to predict only a single sample, it behaves differently than the other threshold metrics which raise exceptions as expected, like roc_auc:

```python
from sklearn.metrics import roc_auc_score

y_score = [1]
y_true = [1]

score = roc_auc_score(y_true=y_true, y_score=y_score)
```

Output:

```
ValueError: Only one class present in y_true. ROC AUC score is not defined in that case.
```

Comment (Innixma) [reply core/tests/unittests/metrics/test_metrics.py:155]: If I were to fix this in AutoGluon, users might complain that AutoGluon's metric raises an exception when scikit-learn's does not.

File: core/tests/unittests/metrics/test_metrics.py (hunk 2 above)
Comment (Innixma) [reply core/tests/unittests/metrics/test_metrics.py:163]: Possibly, however as in the prior comment on average_precision the details are quite nuanced for each, while the scenarios this comes up are very niche.

File: core/tests/unittests/metrics/test_metrics.py (hunk 1 above)
Comment (Innixma) [reply core/tests/unittests/metrics/test_metrics.py:155]: This is an example of the issue in sklearn: https://github.com/scikit-learn/scikit-learn/issues/30147#issuecomment-2438417120

Comment (suzhoum) [reply core/tests/unittests/metrics/test_metrics.py:155]: Thanks for the context!

**Review by suzhoum — COMMENTED** [review]
LGTM!
